### PR TITLE
Add `finality_digest` to `block_refs` in `finality_core` and `reversible_blocks_root` to `finality_digest`

### PR DESCRIPTION
**Pull request by linh2931**
* Add the finality digest associated with a block to each record of the `block_refs` in the `finality_core`. (`block_refs` contain a record for every ancestor block up to and including the last final block).

* For the convenience of finality violation proofs, make the finality digest to commit to the root digest of a Merkle tree over a sequence of records extracted from the sequence of block_refs in the block_refs. This Merkle root digest is called `reversible_blocks_root`. The sequence the Merkle tree commits over is one less than the sequence of block_refs. If there is are no block_refs or only 1 block_ref, then the Merkle root digest can just be the empty digest. Otherwise, the sequence is of records including the block number, finality digest, block timestamp, and the timestamp of the parent block. The `reversible_blocks_root` is added at the same level as current `last_pending_and_base`.
* Convert `last_pending_and_base` into a structure named `combined_base_digsts_t` and replace the long field name `last_pending_finalizer_policy_and_base_digest` in `finality_digest_data_v1` with `combined_base_digsts`. 
* Add `reversible_blocks_root` to SHiP's  `finality_data`.

Resolves https://github.com/AntelopeIO/spring/issues/377

### CODE REVIEW

File: libraries/chain/block_header_state.cpp [hunk 1]
```diff
@@ -38,16 +38,19 @@ digest_type block_header_state::compute_base_digest() const {
...
    finality_digest_data_v1 finality_digest_data {
-      .active_finalizer_policy_generation      = active_finalizer_policy->generation,
-      .final_on_strong_qc_block_num            = core.final_on_strong_qc_block_num,
-      .finality_tree_digest                    = finality_mroot(),
-      .last_pending_finalizer_policy_and_base_digest = lpfp_base_digest
+      .active_finalizer_policy_generation  = active_finalizer_policy->generation,
+      .final_on_strong_qc_block_num        = core.final_on_strong_qc_block_num,
+      .finality_tree_digest                = finality_mroot(),
+      .reversible_blocks_mroot             = core.get_reversible_blocks_mroot(),
+      .combined_base_digsts                = combined_base_digsts
```
Comment (greg7mdp) [inline libraries/chain/block_header_state.cpp:53]:
```suggestion
      .combined_base_digests                = combined_base_digests
```

Comment (linh2931) [reply libraries/chain/block_header_state.cpp]: Thanks. It is replaced by `level_3_commitments`

File: libraries/chain/include/eosio/chain/block_header_state.hpp [hunk 2]
```diff
@@ -20,14 +20,26 @@ namespace detail { struct schedule_info; };
...
+   digest_type base_digest{};
+};
+
+// commitments used in the context of finalizer policy transitions
+struct level_2_commitments_t {
+   digest_type last_pending_fin_pol_digest{};
+   digest_type l3_commitments_digest{};
+};
+
+// finalty digest
```
Comment (greg7mdp) [inline libraries/chain/include/eosio/chain/block_header_state.hpp:35]:
```suggestion
// finality digest
```

Comment (linh2931) [reply libraries/chain/include/eosio/chain/block_header_state.hpp]: Thanks. Fixed.

**Review by greg7mdp — CHANGES_REQUESTED** [review]

File: libraries/chain/block_header_state.cpp [hunk 3]
```diff
@@ -38,16 +38,25 @@ digest_type block_header_state::compute_base_digest() const {
...
    assert(active_finalizer_policy);
    finality_digest_data_v1 finality_digest_data {
-      .active_finalizer_policy_generation      = active_finalizer_policy->generation,
-      .final_on_strong_qc_block_num            = core.final_on_strong_qc_block_num,
-      .finality_tree_digest                    = finality_mroot(),
-      .last_pending_finalizer_policy_and_base_digest = lpfp_base_digest
+      .active_finalizer_policy_generation  = active_finalizer_policy->generation,
+      .final_on_strong_qc_block_num        = core.final_on_strong_qc_block_num,
+      .finality_tree_digest                = finality_mroot(),
+      .l2_commitments_digest               = fc::sha256::hash(level_2_commitments)
```
Comment (greg7mdp) [inline libraries/chain/block_header_state.cpp:59]: Should this be named `l2_finality_digest`?

File: libraries/chain/include/eosio/chain/finality/finality_core.hpp [hunk 4]
```diff
@@ -20,6 +21,14 @@ struct block_ref
    bool operator==(const block_ref& o) const = default;
 };
 
+struct block_ref_digest_data
+{
+   block_num_type   block_num;
```
Comment (heifner) [inline libraries/chain/include/eosio/chain/finality/finality_core.hpp:26]: Init `block_num`

**Review by heifner — CHANGES_REQUESTED** [review]

File: libraries/chain/include/eosio/chain/finality/finality_core.hpp (hunk 4 above)
Comment (linh2931) [reply libraries/chain/include/eosio/chain/finality/finality_core.hpp]: Done.

File: libraries/chain/block_header_state.cpp (hunk 3 above)
Comment (linh2931) [reply libraries/chain/block_header_state.cpp:59]: It was decided not change the name.

File: libraries/chain/finality/finality_core.cpp [hunk 5]
```diff
@@ -148,6 +149,34 @@ const block_ref& finality_core::get_block_reference(block_num_type block_num) co
...
+ */
+digest_type finality_core::get_reversible_blocks_mroot() const {
+   // Parent block timestamp is not availale as parent block does not exist
+   if (refs.size() <= 1) {
+      return {};
+   }
+
+   // Build a merkle tree of a sequence of records including block number,
+   // block timestamp, finality digest, and the timestamp of the parent block.
+   deque<digest_type> block_ref_digests;
```
[the remark this answers is not in the document]
Comment (linh2931) [reply libraries/chain/finality/finality_core.cpp]: Done

**Review by greg7mdp — APPROVED** [review]

**Review by heifner — APPROVED** [review]

**Comment by ericpassmore** [comment]
Note:start
group: STABILITY
category: INTERNALS
summary: For the convenience of finality violation proofs, make the finality digest to commit to the root digest of a Merkle tree over a sequence of records extracted from the sequence of block_refs in the block_refs. This Merkle root digest is called reversible_blocks_root.
Note:end

**Pull request merged**